In [2]:
import pandas as pd 

## 1. Loading Data

In [6]:
jan_taxi = pd.read_parquet('./data/yellow_tripdata_2023-01.parquet')
jan_taxi

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
0               2  2023-01-01 00:32:10   2023-01-01 00:40:36              1.0   
1               2  2023-01-01 00:55:08   2023-01-01 01:01:27              1.0   
2               2  2023-01-01 00:25:04   2023-01-01 00:37:49              1.0   
3               1  2023-01-01 00:03:48   2023-01-01 00:13:25              0.0   
4               2  2023-01-01 00:10:29   2023-01-01 00:21:19              1.0   
...           ...                  ...                   ...              ...   
3066761         2  2023-01-31 23:58:34   2023-02-01 00:12:33              NaN   
3066762         2  2023-01-31 23:31:09   2023-01-31 23:50:36              NaN   
3066763         2  2023-01-31 23:01:05   2023-01-31 23:25:36              NaN   
3066764         2  2023-01-31 23:40:00   2023-01-31 23:53:00              NaN   
3066765         2  2023-01-31 23:07:32   2023-01-31 23:21:56              NaN   

         trip_distance  RatecodeID store_and_fwd_flag  PULocationID  \
0                 0.97         1.0                  N           161   
1                 1.10         1.0                  N            43   
2                 2.51         1.0                  N            48   
3                 1.90         1.0                  N           138   
4                 1.43         1.0                  N           107   
...                ...         ...                ...           ...   
3066761           3.05         NaN               None           107   
3066762           5.80         NaN               None           112   
3066763           4.67         NaN               None           114   
3066764           3.15         NaN               None           230   
3066765           2.85         NaN               None           262   

         DOLocationID  payment_type  fare_amount  extra  mta_tax  tip_amount  \
0                 141             2         9.30   1.00      0.5        0.00   
1                 237             1         7.90   1.00      0.5        4.00   
2                 238             1        14.90   1.00      0.5       15.00   
3                   7             1        12.10   7.25      0.5        0.00   
4                  79             1        11.40   1.00      0.5        3.28   
...               ...           ...          ...    ...      ...         ...   
3066761            48             0        15.80   0.00      0.5        3.96   
3066762            75             0        22.43   0.00      0.5        2.64   
3066763           239             0        17.61   0.00      0.5        5.32   
3066764            79             0        18.15   0.00      0.5        4.43   
3066765           143             0        15.97   0.00      0.5        2.00   

         tolls_amount  improvement_surcharge  total_amount  \
0                 0.0                    1.0         14.30   
1                 0.0                    1.0         16.90   
2                 0.0                    1.0         34.90   
3                 0.0                    1.0         20.85   
4                 0.0                    1.0         19.68   
...               ...                    ...           ...   
3066761           0.0                    1.0         23.76   
3066762           0.0                    1.0         29.07   
3066763           0.0                    1.0         26.93   
3066764           0.0                    1.0         26.58   
3066765           0.0                    1.0         21.97   

         congestion_surcharge  airport_fee  
0                         2.5         0.00  
1                         2.5         0.00  
2                         2.5         0.00  
3                         0.0         1.25  
4                         2.5         0.00  
...                       ...          ...  
3066761                   NaN          NaN  
3066762                   NaN          NaN  
3066763                   NaN          NaN  
3066764                   N

In [7]:
feb_taxi = pd.read_parquet('./data/yellow_tripdata_2023-02.parquet')
feb_taxi

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
0               1  2023-02-01 00:32:53   2023-02-01 00:34:34              2.0   
1               2  2023-02-01 00:35:16   2023-02-01 00:35:30              1.0   
2               2  2023-02-01 00:35:16   2023-02-01 00:35:30              1.0   
3               1  2023-02-01 00:29:33   2023-02-01 01:01:38              0.0   
4               2  2023-02-01 00:12:28   2023-02-01 00:25:46              1.0   
...           ...                  ...                   ...              ...   
2913950         2  2023-02-28 23:46:00   2023-03-01 00:05:00              NaN   
2913951         2  2023-02-28 23:26:02   2023-02-28 23:37:10              NaN   
2913952         2  2023-02-28 23:24:00   2023-02-28 23:38:00              NaN   
2913953         2  2023-02-28 23:03:00   2023-02-28 23:10:00              NaN   
2913954         2  2023-02-28 23:03:03   2023-02-28 23:12:51              NaN   

         trip_distance  RatecodeID store_and_fwd_flag  PULocationID  \
0                 0.30         1.0                  N           142   
1                 0.00         1.0                  N            71   
2                 0.00         1.0                  N            71   
3                18.80         1.0                  N           132   
4                 3.22         1.0                  N           161   
...                ...         ...                ...           ...   
2913950           4.65         NaN               None           249   
2913951           2.47         NaN               None           186   
2913952           3.49         NaN               None           158   
2913953           2.13         NaN               None            79   
2913954           2.28         NaN               None           161   

         DOLocationID  payment_type  fare_amount  extra  mta_tax  tip_amount  \
0                 163             2         4.40   3.50      0.5        0.00   
1                  71             4        -3.00  -1.00     -0.5        0.00   
2                  71             4         3.00   1.00      0.5        0.00   
3                  26             1        70.90   2.25      0.5        0.00   
4                 145             1        17.00   1.00      0.5        3.30   
...               ...           ...          ...    ...      ...         ...   
2913950           140             0        20.22   0.00      0.5        4.84   
2913951            79             0        13.66   0.00      0.5        2.65   
2913952           143             0        17.64   0.00      0.5        0.00   
2913953           162             0        13.56   0.00      0.5        2.63   
2913954           140             0        14.89   0.00      0.5        3.78   

         tolls_amount  improvement_surcharge  total_amount  \
0                 0.0                    1.0          9.40   
1                 0.0                   -1.0         -5.50   
2                 0.0                    1.0          5.50   
3                 0.0                    1.0         74.65   
4                 0.0                    1.0         25.30   
...               ...                    ...           ...   
2913950           0.0                    1.0         29.06   
2913951           0.0                    1.0         20.31   
2913952           0.0                    1.0         21.64   
2913953           0.0                    1.0         20.19   
2913954           0.0                    1.0         22.67   

         congestion_surcharge  Airport_fee  
0                         2.5         0.00  
1                         0.0         0.00  
2                         0.0         0.00  
3                         0.0         1.25  
4                         2.5         0.00  
...                       ...          ...  
2913950                   NaN          NaN  
2913951                   NaN          NaN  
2913952                   NaN          NaN  
2913953                   N

### How many columns in the  data for January?
    There is 19 

## 2. Computing duration
### What's the standard deviation of the trips duration in January?

In [20]:
jan_taxi['duration'] = abs(jan_taxi['tpep_pickup_datetime']-jan_taxi['tpep_dropoff_datetime'])
jan_taxi['duration_minute'] = jan_taxi['duration'].dt.total_seconds()/60
jan_taxi['duration_minute']

0           8.433333
1           6.316667
2          12.750000
3           9.616667
4          10.833333
             ...    
3066761    13.983333
3066762    19.450000
3066763    24.516667
3066764    13.000000
3066765    14.400000
Name: duration_minute, Length: 3066766, dtype: float64

In [27]:
jan_taxi['duration_minute'].std()

42.59433614393127

The standard deviation of trips duration in january is 42.59

## 3. Dropping outliers
#### Let's remove them and keep only the records where the duration was between 1 and 60 minutes (inclusive)

In [29]:
jan_taxi = jan_taxi[(jan_taxi['duration_minute']>=1) & (jan_taxi['duration_minute']<=60)]
jan_taxi

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
0               2  2023-01-01 00:32:10   2023-01-01 00:40:36              1.0   
1               2  2023-01-01 00:55:08   2023-01-01 01:01:27              1.0   
2               2  2023-01-01 00:25:04   2023-01-01 00:37:49              1.0   
3               1  2023-01-01 00:03:48   2023-01-01 00:13:25              0.0   
4               2  2023-01-01 00:10:29   2023-01-01 00:21:19              1.0   
...           ...                  ...                   ...              ...   
3066761         2  2023-01-31 23:58:34   2023-02-01 00:12:33              NaN   
3066762         2  2023-01-31 23:31:09   2023-01-31 23:50:36              NaN   
3066763         2  2023-01-31 23:01:05   2023-01-31 23:25:36              NaN   
3066764         2  2023-01-31 23:40:00   2023-01-31 23:53:00              NaN   
3066765         2  2023-01-31 23:07:32   2023-01-31 23:21:56              NaN   

         trip_distance  RatecodeID store_and_fwd_flag  PULocationID  \
0                 0.97         1.0                  N           161   
1                 1.10         1.0                  N            43   
2                 2.51         1.0                  N            48   
3                 1.90         1.0                  N           138   
4                 1.43         1.0                  N           107   
...                ...         ...                ...           ...   
3066761           3.05         NaN               None           107   
3066762           5.80         NaN               None           112   
3066763           4.67         NaN               None           114   
3066764           3.15         NaN               None           230   
3066765           2.85         NaN               None           262   

         DOLocationID  payment_type  ...  extra  mta_tax  tip_amount  \
0                 141             2  ...   1.00      0.5        0.00   
1                 237             1  ...   1.00      0.5        4.00   
2                 238             1  ...   1.00      0.5       15.00   
3                   7             1  ...   7.25      0.5        0.00   
4                  79             1  ...   1.00      0.5        3.28   
...               ...           ...  ...    ...      ...         ...   
3066761            48             0  ...   0.00      0.5        3.96   
3066762            75             0  ...   0.00      0.5        2.64   
3066763           239             0  ...   0.00      0.5        5.32   
3066764            79             0  ...   0.00      0.5        4.43   
3066765           143             0  ...   0.00      0.5        2.00   

         tolls_amount  improvement_surcharge  total_amount  \
0                 0.0                    1.0         14.30   
1                 0.0                    1.0         16.90   
2                 0.0                    1.0         34.90   
3                 0.0                    1.0         20.85   
4                 0.0                    1.0         19.68   
...               ...                    ...           ...   
3066761           0.0                    1.0         23.76   
3066762           0.0                    1.0         29.07   
3066763           0.0                    1.0         26.93   
3066764           0.0                    1.0         26.58   
3066765           0.0                    1.0         21.97   

         congestion_surcharge  airport_fee        duration duration_minute  
0                         2.5         0.00 0 days 00:08:26        8.433333  
1                         2.5         0.00 0 days 00:06:19        6.316667  
2                         2.5         0.00 0 days 00:12:45       12.750000  
3                         0.0         1.25 0 days 00:09:37        9.616667  
4                         2.5         0.00 0 days 00:10:50       10.833333  
...                       ...          ...             ...             ...  
3066761                   NaN     

In [31]:
len(jan_taxi)

3009176

In [33]:
percentage_left = (3009176/3066766)*100
percentage_left 

98.12212604417813

After removing outliers, we have left 98%

## 4. One Hot Encoding 
#### Let's apply one-hot encoding to the pickup and dropoff location IDs. We'll use only these two features for our model.

>- Turn the dataframe into a list of dictionaries (remember to re-cast the ids to strings otherwise it will label encode them)
>- Fit a dictionary vectorizer
>- Get a feature matrix from it

In [ ]:
# re-cast VendorId to strings
jan_taxi['VendorID'].astype('str')
# Turn the dataframe into a list od dictionnaries
records = jan_taxi.to_dict(orient="records")